File: TiaCodegen/Commands/Coils/BaseCoil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TiaCodegen.Commands.Signals;
using TiaCodegen.Interfaces;

namespace TiaCodegen.Commands.Coils
{
    public class BaseCoil : IOperationOrSignal, ICoil
    {
        public string DebugInfo { get; set; }
        public BaseCoil(Signal signal, IOperationOrSignal op = null)
        {
            Signal = signal;
            Children = new List<IOperationOrSignal>();
            Children.Add(signal);
            if (op != null)
                Children.Add(op);
        }

        public bool Negated { get; set; }

        public Signal Signal { get; set; }

        public virtual List<IOperationOrSignal> Children { get; set; }

        public ulong OperationId { get; set; }

        public IOperationOrSignal Parent { get; set; }

        public bool DoNotCreateContact { get; set; }

        public int Cardinality { get; set; }

        public virtual int CreateContactAndFillCardinality(IOperationOrSignal parent)
        {
            this.Cardinality = 1;
            return this.Cardinality;
        }

        public override string ToString()
        {
            return this.GetType().Name;
        }

        public IOperationOrSignal Clone()
        {
            try
            {
                var inst = (BaseOperationOrSignal)Activator.CreateInstance(this.GetType(), this.Signal.Clone());
                var props = this.GetType().GetProperties();
                foreach (var p in props)
                {
                    if (p.Name == "Children" || p.Name == "Parent" || p.Name == "SignalId" || p.Name == "OperationId" || p.Name == "Cardinality" || p.Name == "DoNotCreateContact")
                        continue;
                    p.SetValue(inst, p.GetValue(this));
                }
                foreach (var c in Children.Skip(1))
                {
                    inst.Children.Add(c.Clone());
                }

                return inst;
            }
            catch (Exception ex)
            {
                throw new Exception("Error Clone: " + this.GetType().Name, ex);
            }
        }
    }
}

File: TiaCodegen-ts/src/Commands/Coils/BaseCoil.ts
import { IOperationOrSignal } from '../../Interfaces/IOperationOrSignal';
import { ICoil } from '../../Interfaces/ICoil';
import { Signal } from '../Signals/Signal';

export class BaseCoil implements IOperationOrSignal, ICoil {
    debugInfo: string | null = null;
    negated: boolean = false;
    signal: Signal;
    children: IOperationOrSignal[];
    operationId: number = 0;
    parent: IOperationOrSignal | null = null;
    doNotCreateContact: boolean = false;
    cardinality: number = 0;

    constructor(signal: Signal, op: IOperationOrSignal | null = null) {
        this.signal = signal;
        this.children = [];
        this.children.push(signal);
        if (op !== null) {
            this.children.push(op);
        }
    }

    createContactAndFillCardinality(_parent: IOperationOrSignal): number {
        this.cardinality = 1;
        return 1;
    }

    toString(): string {
        return this.constructor.name;
    }

    clone(): IOperationOrSignal {
        const inst = new (this.constructor as new (s: Signal) => BaseCoil)(this.signal.clone() as Signal);
        inst.debugInfo = this.debugInfo;
        inst.negated = this.negated;
        for (const c of this.children.slice(1)) {
            inst.children.push(c.clone());
        }
        return inst;
    }
}
